pkg/adapter/reconciler/blueprint_controller.go

```diff
@@ -39,28 +39,27 @@ func (r *BlueprintReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 
 	err := r.blueprintChangeHandler.HandleChange(ctx, req.Name)
 
-	var internalError *domainservice.InternalError
-	var conflictError *domainservice.ConflictError
-	var notFoundError *domainservice.NotFoundError
-	var invalidError *domain.InvalidBlueprintError
 	if err != nil {
 		logger := logger.WithValues("error", err)
-		if errors.As(err, &internalError) {
+		switch err.(type) {
+		case *domainservice.InternalError:
 			logger.Error(err, "An internal error occurred and can maybe be fixed by retrying it later")
 			return ctrl.Result{}, err //automatic requeue
-		} else if errors.As(err, &conflictError) {
+		case *domainservice.ConflictError:
 			logger.Info("A concurrent update happened in conflict to the processing of the blueprint spec. A retry could fix this issue")
 			return ctrl.Result{Requeue: true, RequeueAfter: 1 * time.Second}, nil // no error as this would lead to the ignorance of our own retry params
-		} else if errors.As(err, &notFoundError) {
+		case *domainservice.NotFoundError:
 			logger.Info("blueprint was not found, so maybe it was deleted in the meantime. No further evaluation will happen")
 			return ctrl.Result{Requeue: false}, nil
-		} else if errors.As(err, &invalidError) {
+		case *domain.InvalidBlueprintError:
 			logger.Info("blueprint is invalid, therefore there will be no further evaluation.")
 			return ctrl.Result{Requeue: false}, nil
+		default:
+			logger.Error(err, "an unknown error type occurred. Retry with default backoff")
+			return ctrl.Result{}, err //automatic requeue
 		}
-		logger.Error(err, "an unknown error type occurred. Retry with default backoff")
-		return ctrl.Result{}, err //automatic requeue
 	}
+
 	return ctrl.Result{}, nil
 }
```

Comment:
i think, this will not work as expected as a direct type cast cannot detect the type of wrapped errors but errors.As on the other hand considers all wrapped errors. In this operator the errors will usually be wrapped.

Reply:
Thanks for the comment, now I see the problem. Anyhow the if-else-branch-o-rama is still quite ugly in my eyes because if-else statements are rather unreadable compared to other constructs.

maybe I can come up with a working solution which avoids large (and potentially growing) if-else lists.

Reply:
Yeah, I guess you could write a recusive function that tests against the `switch`, and if it doesn't match calls itself with the unwrapped error. If the error it is called with is `nil`, we know we have reached the end of the unwrap chain and must return early.

Reply:
Although that would make it ugly to retain the original error value for further processing.

Reply:
I reworked the solution so it actually catches the desired error in the error chain. It more or less looks like before with the difference that it was extracted into a function which uses switch-case instead of if-else branches.